Sortedcontainerupdates

StephenVavasis opened the pull request:
Major update to sorted containers.

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/239: Dict constructors on typed data but without explicit type declarations should no longer return `{Any,Any}` items because collect is now used when explicit types are omitted

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/479: Deprecate `insert!`, `pop!(s)`, introduce `popfirst!`, `poplast!`, `ss_push!`, `sd_push!`, `smd_push!`

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/528: `push!` documentation improved

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/667: Docs extensively cleaned up and much greater use of documenter.jl

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/669: Expanded use of tokens including `a:b` notation for tokens

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/671: Add reverse iteration

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/742: New union, merge, symdiff, setdiff, intersect algorithms for sorted containers to improve performance

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/726: Clarify docs on `ss_push!`, `sd_push!`, and `sdm_push!` (formerly `insert!`)

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/767: Implement firstindex, lastindex, reverse for sorted container iteration

Address issue https://github.com/JuliaCollections/DataStructures.jl/issues/781: Use collect on an intermediate step
to infer types in untyped construction of sorted containers

Also, this commit cleans up the constructors and adds two new iteration-helper functions `inclusive_key` and `exclusive_key` that were useful in my own work.

StephenVavasis commented:
The following test from the new testset, which is meant to exercise a new `show` method, is failing on some platforms:
```julia
    my_assert(remove_spaces(repr(MIME{Symbol("text/plain")}(), factors5)) ==
              "SortedMultiDict{Char,Int64,ForwardOrdering}" *
              "(ForwardOrdering(),['a'=>1,'b'=>2,'c'=>3])")
```
Here, `remove_spaces` is a locally written function-- see below.  So apparently on some platforms, `show` works differently?  Anyone know the differences?  Or is there a better way to write a unit test for `show`?

```julia
function remove_spaces(s::String)
    b = Vector{UInt8}()
    for c in s
        if !isspace(c)
            push!(b,UInt8(c))
        end
    end
    String(b)
end
```

StephenVavasis commented:
The reason the test is failing is: In Julia 1.0, printing the vector `[1=>'a',2=>'b']` prepends the output with `Pair{Int64,Char}`.  Prepending this eltype-declaration does not happen in Julia 1.7.  Apparently I need to think of a more robust unit test for a `show` method.

StephenVavasis commented:
I replaced the failing unit test for `show` that does not require precisely a particular string, and now all tests are successful.

oxinabox reviewed (commented):
sorry I have not finished reviewing this.
This broadly speaking looks good and if all the existing tests still pass then it should be good

StephenVavasis commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+
+const SDMContainer = Union{SortedDict, SortedMultiDict}
+const SortedContainer = Union{SDMContainer, SortedSet}
+const Token = Tuple{SortedContainer, IntSemiToken}
```
Are you asking why `Token` is implemented as a tuple rather than a `struct`?   This is a legacy issue.  In pre-1.0 Julia, any struct that contained an immutable object had itself to be heap allocated.  However, in pre-1.0 Julia, operations on tuples containing an immutable object didn't always allocate.  In particular, `deref((mysorteddict, semitok))` did not heap-allocate its 2-tuple argument, whereas a hypothetical `deref(Token(mysorteddict, semitok))`, where `Token` is the constructor of a hypothetical struct, would have allocated.

oxinabox replied:
More my surprise was that a token included both the container and the indexing object.
I would have expected the token to just be the indexing object (but that is defined as a semitoken)
is the idea that the token is enough to fully retrieve the stored object?
But semitoken is only half of what is expected?

StephenVavasis replied:
I think full tokens are useful, e.g., `deref(token_firstindex(s))` as opposed to `deref((s,firstindex(s)))` where the user needs to be consistent with the argument `s`.  So if it is OK with you, I would like to leave these.

oxinabox commented on src/DataStructures.jl:
```diff
@@ -60,6 +65,14 @@ module DataStructures
+    import Base.+
+    import Base.-
+    import Base.isequal
+    export sd_push!, ss_push!, smd_push!, poplast!
```
Rather than all 3 `sd_push!, ss_push!, smd_push!` is there a single name we can overload for these that refers to what it returns, rather than hungarian notation for it's first argument?
Like `push_and_get_token!` or something? (is that what they are doing?)

StephenVavasis replied:
Done.  The new function is called `push_return_token!`.  Since it is a single function name, I unified the APIs by making the return variables `Bool,IntSemiToken`. The `Bool` return variable is redundant for SortedMultiDict because push always creates a new record for SortedMultiDict (i.e., the first return variable is always `true` for a SortedMultiDict).

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+the first entry of the sorted container `m`, or the past-end token
+if the container is empty.  Time: O(log *n*)
+"""  
+token_firstindex(m::SortedContainer) = (m, firstindex(m))
```
do we need this to be exported?
It seems like if someone needs it is is a trivial and obvious definition.

StephenVavasis replied:
See other comment about `token_firstindex`.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+if the container is empty.  Time: O(log *n*)
+"""
+Base.lastindex(m::SortedContainer) = endof(m)
+endof(m::SortedContainer) = IntSemiToken(endloc(m.bt))
```
as per `startof` do we need this? or can we just use `lastindex`?

StephenVavasis replied:
I have deprecated `startof` and `endof` in favor of `firstindex` and `lastindex`.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+last entry of the sorted container `m`, or the before-start semitoken
+if the container is empty.  Time: O(log *n*)
+"""
+token_lastindex(m::SortedContainer) = (m, lastindex(m))
```
as per `token_firstindex` do we need to export this?
It is a trivial and obvious definition

StephenVavasis replied:
See other comment about `token_firstindex`

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+    return Pair(kdrec.k, kdrec.d)
+end
+
+function deref(ii::SortedDictToken)
+    has_data(ii)
+    deref_nocheck(ii)
+end
+
+@inline function deref_nocheck(ii::SortedMultiDictToken)
+    @inbounds kdrec = ii[1].bt.data[ii[2].address]
+    return kdrec.k => kdrec.d
+end
```
similar for the other `foo`/`foo_nocheck` pairs?

StephenVavasis replied:
I tried to implement this, but it is not working properly, so perhaps I am misunderstanding the rules.  Here is the new definition of `deref`:

```julia
Base.@propagate_inbounds function deref(ii::SortedDictToken)
    Base.@boundscheck has_data(ii)
    @inbounds kdrec = ii[1].bt.data[ii[2].address]
    return Pair(kdrec.k, kdrec.d)
end
```

My intention is that if a `deref` operation is prefixed by `@inbounds`, then `has_data` will not be invoked.  But this is not happening:

```julia
julia> s = SortedDict{Int,String}()
SortedDict{Int64, String, Base.Order.ForwardOrdering}()

julia> deref(token_firstindex(s))
ERROR: BoundsError
Stacktrace:
 [1] has_data
   @ c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:442 [inlined]
 [2] deref(ii::Tuple{SortedDict{Int64, String, Base.Order.ForwardOrdering}, DataStructures.Tokens.IntSemiToken})
   @ DataStructures c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:242
 [3] top-level scope
   @ REPL[24]:1

julia> @inbounds deref(token_firstindex(s))
ERROR: BoundsError
Stacktrace:
 [1] has_data
   @ c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:442 [inlined]
 [2] deref(ii::Tuple{SortedDict{Int64, String, Base.Order.ForwardOrdering}, DataStructures.Tokens.IntSemiToken})
   @ DataStructures c:\Users\vavasis\OneDrive - University of Waterloo\Julia_pkg_devdir\DataStructures.jl\src\sorted_container_iteration.jl:242
 [3] top-level scope
   @ REPL[25]:1
```

StephenVavasis replied:
OK, nvm, I figured this out too.  Apparently `@inbounds` is not necessarily active if it is called directly from the REPL as opposed to its being called from a compiled function.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+
+These five invocations of `in` use the index structure of the sorted
+container and test equality based on the order object of the keys
+rather than `isequal`. Therefore, these five are all faster than
```
should we comment here that if `isequal` is defined consistently with `isless` etc this should have same behavour?

StephenVavasis replied:
The issue here is not whether `isless` and `isequal` are consistent.  Rather, the issue is whether the user has defined an ordering (such as case-insensitive for strings) for the container different from `isequal` and `isless`.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+
+(:)(t1::Token, t2::Token) = _colon(t1,t2)
+
+function _colon(t1::Token, t2::Token)
```
why not define this as `:` ?

StephenVavasis replied:
The definition `function Base.(:)(a,b)` gives a syntax error (noted above).  However, `function (:)(a,b)` works, so I have fixed this (but I still have the issue with `import` noted above).

StephenVavasis replied:
OK, I figured this out.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+then the before-start/past-end tokens are returned (and there is no error).
+Time: O(*j*+log *n*), so this function is not optimized for long jumps.
+"""
++(t1::Token, numstep::Integer) = 
+    numstep >= 0 ? stepforward(t1, numstep) : stepback(t1, -numstep)
+
```
this easily fits on 1 line, no?
```suggestion
+(t1::Token, numstep::Integer) = numstep >= 0 ? stepforward(t1, numstep) : stepback(t1, -numstep)

```

StephenVavasis replied:
Done.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
++(t1::Token, numstep::Integer) = 
+    numstep >= 0 ? stepforward(t1, numstep) : stepback(t1, -numstep)
+
+-(t1::Token, numstep::Integer) = 
+    numstep >= 0 ? stepback(t1, numstep) : stepforward(t1, -numstep)
```
this easily fits on 1 line, no?
```suggestion
-(t1::Token, numstep::Integer) = numstep >= 0 ? stepback(t1, numstep) : stepforward(t1, -numstep)
```

StephenVavasis replied:
Done.

oxinabox commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+function stepforward(t1::Token, numstep::Integer)
+    m = t1[1]
+    j = t1[2].address
+    !(j in m.bt.useddatacells) && throw(BoundsError())
```
what do you think of
```suggestion
    (j in m.bt.useddatacells) || throw(BoundsError())
```

StephenVavasis replied:
I can make this change, but I have used the idiom `!(correct) && throw(...)` (rather than `correct || throw(...)`) throughout my code.  Do you recommend that I change them all.

oxinabox commented on src/sorted_dict.jl:
```diff
@@ -3,411 +3,294 @@
-
-"""
-    in(p, sc)
+@deprecate insert!(m::SortedDict, k, d) sd_push!(m::SortedDict, k=>d)
```
can we move this into `src/deprecations.jl` so we remember to delete it in next release?

StephenVavasis replied:
Done

oxinabox commented on test/test_sorted_containers.jl:
```diff
@@ -1,18 +1,24 @@
-import Base.Forward
-import Base.Reverse
-import DataStructures.eq
+using Base: Ordering, Forward, Reverse, ForwardOrdering,
+            ReverseOrdering
```
this fits on one line
```suggestion
using Base: Ordering, Forward, Reverse, ForwardOrdering, ReverseOrdering
```

StephenVavasis replied:
Done

oxinabox commented on test/test_sorted_containers.jl:
```diff
@@ -1,18 +1,24 @@
 import Base.lt
-import Base.ForwardOrdering
-import Base.ReverseOrdering
-import DataStructures.IntSemiToken
+import DataStructures.eq
```
We should make the imports into `using`s i think

StephenVavasis replied:
Done

oxinabox commented on test/test_sorted_containers.jl:
```diff
@@ -1,18 +1,24 @@
+using Base: Ordering, Forward, Reverse, ForwardOrdering,
+            ReverseOrdering
+using DataStructures: IntSemiToken
 import Base.lt
```
isn't this called `Base.lessthan` now days?

StephenVavasis replied:
I checked `ordering.jl` in `Base`, and I still see `lt`.  Where did you see `lessthan`?

oxinabox commented on test/test_sorted_containers.jl:
```diff
@@ -1,18 +1,24 @@
 lt(::CaseInsensitive, a, b) = isless(lowercase(a), lowercase(b))
 eq(::CaseInsensitive, a, b) = isequal(lowercase(a), lowercase(b))
 
+
```
```suggestion
# for testing custom ordering
```

oxinabox commented on test/test_sorted_containers.jl:
```diff
@@ -32,6 +38,17 @@ function my_primes(N)
     p
 end
 
+function remove_spaces(s::String)
+    b = Vector{UInt8}()
+    for c in s
+        if !isspace(c)
+            push!(b,UInt8(c))
+        end
+    end
+    String(b)
+end
+    
```
```suggestion
remove_spaces(s::String) = replace(s, re"\s+"=>"")    
```

StephenVavasis replied:
I found out that it is actually `r"\s+"` rather than `re"\s+"`, but aside from that, this is done.

oxinabox commented on test/test_sorted_containers.jl:
```diff
@@ -1765,3 +2093,4 @@ end
         end
     end
 end
+
```
can we label the new tests with a comment including the URL of the github issues that they prove is solved?

oxinabox reviewed (commented):
Nice work.
I assume that the logic is at least as correct as before as all previous tests still pass.

Some mostly stylistic comments to resolve

StephenVavasis commented on src/DataStructures.jl:
```diff
@@ -60,6 +65,14 @@ module DataStructures
 
     export findkey
 
+    import Base.==
+    import Base.(:)
+    import Base.+
+    import Base.-
+    import Base.isequal
```
I am having trouble with this for `==` and `:`.  In particular, the following constructs give me syntax errors:
```
Base.(==)(a,b) = ...
Base.==(a,b) = ...
function Base.(:)(a,b)
```
whereas there are no syntax errors with
```
import Base.==
import Base.(:)

==(a,b) = ...
function (:)(a,b)
```
Any thoughts how to fix this?

StephenVavasis replied:
I figured this out.

StephenVavasis commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+    token_firstindex(m::SortedContainer)
+
+
+Return the token of
```
I would like to leave these.  See other comment.

StephenVavasis commented on src/sorted_container_iteration.jl:
```diff
@@ -0,0 +1,1167 @@
+sorted container `m` that is greater than or equal to
+`k` in the sort order.
+If there is no
+such item, then the past-end semitoken is returned.  Time: O(*c* log *n*)
```
I didn't know this!  However, if you don't mind, I'll leave it as it is because there is not much benefit in changing it compared to the effort needed.

StephenVavasis commented on docs/Project.toml:
```diff
@@ -1,5 +1,4 @@
 Documenter = "e30172f5-a6a5-5a46-863b-614d45cd2de4"
-
-[compat]
-Documenter = "0.23"
```
OK, I put it back.  I'm not sure how it got deleted.

StephenVavasis commented on docs/src/sorted_containers.md:
```diff
@@ -10,18 +10,19 @@ with the additional feature that the keys are stored in sorted order and
 
 The containers internally use a 2-3 tree, which is a kind of balanced
-tree and is described in many elementary data structure textbooks.
+tree and is described in data structure textbooks.
```
done

StephenVavasis commented on docs/src/sorted_containers.md:
```diff
@@ -85,616 +61,378 @@ token can be regressed. A dereferencing operation on either leads to an
+is not a bits-type.
+If code profiling indicates that statements using tokens are allocating memory,
+then it may be advisable to rewrite the application code using semitokens
+more than tokens.
```
done

StephenVavasis commented on docs/src/sorted_containers.md:
````diff
@@ -85,616 +61,378 @@ token can be regressed. A dereferencing operation on either leads to an
-    searchsortedfirst(sc,k)
+```@docs
+Base.first(sc::SortedContainer)
+```
````
I didn't know about the other syntax for documenter, but now that you told me, I have fixed it.

StephenVavasis commented on docs/src/sorted_containers.md:
````diff
@@ -85,616 +61,378 @@ token can be regressed. A dereferencing operation on either leads to an
-pop!(sc::SortedDict, k)
+ss_push!(ss::SortedSet, k)
 ```
 
 ```@docs
-pop!(ss::SortedSet, k)
+sd_push!(sd::SortedDict, p::Pair)
 ```
 
 ```@docs
-pop!(ss::SortedSet)
+smd_push!(smd::SortedMultiDict, p::Pair)
 ```
````
Done

StephenVavasis commented:
@oxinabox I believe I have addressed all of your comments in this PR.  In the last commit that I just made, I have added comments to the test sets to indicate where the various issues are tested.  Let me know if you need anything else before you can merge this PR.

oxinabox commented on src/DataStructures.jl:
```diff
@@ -39,12 +42,14 @@ module DataStructures
     export SDToken, SDSemiToken, SMDToken, SMDSemiToken
     export SetToken, SetSemiToken
-    export startof
+    export startof, endof ## both are deprecated
```
We don't need to explictly export deprecated things `@deprecate` does that for us

```suggestion
```

oxinabox reviewed (approved):
Sounds good.
You have merge rights.
Merge when you are happy
